src/app/PurchasePanel.tsx

```diff
@@ -0,0 +1,205 @@
+  ) {
+    stateFgColor = "text-green-500";
+    stateBgColor = "bg-green-50";
+    stateDescription = "You are ready to purchase";
+  } else if (
+    prePurchaseCheckState.FailureReason ==
+    PrePurchaseFailureReason.MAX_WALLETS_USED
+  ) {
+    stateFgColor = "text-amber-500";
+    stateBgColor = "bg-amber-50";
+    stateDescription =
+      "Maximum number of wallets reached. This entity cannot use the connected wallet to commit funds to this sale. Please use a previous wallet or select a different investing entity.";
+  } else if (
+    prePurchaseCheckState.FailureReason ==
+    PrePurchaseFailureReason.NO_RESERVED_ALLOCATION
+  ) {
+    stateFgColor = "text-amber-500";
+    stateBgColor = "bg-amber-50";
+    stateDescription =
+      "No reserved allocation. The connected wallet has no reserved allocation for this sale. Please connect a different wallet.";
+  } else {
+    stateFgColor = "text-red-500";
+    stateBgColor = "bg-red-50";
+    stateDescription = "Unknown failure reason. Please contact support.";
+  }
+
+  return (
+    <div className={`${stateBgColor} p-2 rounded-md w-full`}>
+      <p className={`${stateFgColor} w-full`}>{stateDescription}</p>
+    </div>
+  );
+}
+
+function PurchasePanel() {
+  const { address, isConnected } = useAccount();
+  const {
+    authenticated,
+    loading: entityLoading,
+    entity,
+  } = useSonarEntity({
+    saleUUID: sonarConfig.saleUUID,
+    wallet: { address, isConnected },
+  });
+  const client = useSonarClient();
+
+  const [successMessage, setSuccessMessage] = useState<string | null>(null);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+
+  const [prePurchaseCheckState, setPrePurchaseCheckState] = useState<{
+    loading: boolean;
+    value?: PrePurchaseCheckResponse;
+    error?: Error;
+  }>({
+    loading: false,
+  });
+
+  useEffect(() => {
+    const prePurchaseCheck = async () => {
+      if (!address || !authenticated || !entity) {
+        return;
+      }
+
+      setPrePurchaseCheckState((prev) => ({
+        ...prev,
+        loading: true,
+        error: undefined,
+      }));
+
+      try {
+        const response = await client.prePurchaseCheck({
+          saleUUID: sonarConfig.saleUUID,
+          entityType: entity.EntityType,
+          entityUUID: entity.EntityUUID,
+          walletAddress: address,
+        });
+        setPrePurchaseCheckState({
+          loading: false,
+          value: response,
+        });
+      } catch (error) {
+        setPrePurchaseCheckState({
+          loading: false,
+          error: error instanceof Error ? error : new Error(String(error)),
+        });
+      }
+    };
+
+    prePurchaseCheck();
+  }, [address, authenticated, entity, client]);
+
+  if (
+    !address ||
+    entityLoading ||
+    entity?.SaleEligibility !== SaleEligibility.ELIGIBLE ||
+    entity?.EntitySetupState !== EntitySetupState.COMPLETE ||
+    !prePurchaseCheckState
+  ) {
```

Comment:
might want to handle them differently, e.g. `entityLoading` should probably get a loading screen